nion/swift/Inspector.py

```diff
@@ -918,245 +926,246 @@ def _set_property_value(self, property_value: typing.Any) -> None:
+
+        self.data_row_model = DisplayLayerPropertyCommandModel(document_controller, display_item, display_layer, "data_row")
+        self.fill_color_model = DisplayLayerPropertyCommandModel(document_controller, display_item, display_layer, "fill_color")
+        self.stroke_color_model = DisplayLayerPropertyCommandModel(document_controller, display_item, display_layer, "stroke_color")
+        self.stroke_width_model = DisplayLayerPropertyCommandModel(document_controller, display_item, display_layer, "stroke_width")
+
+        self.show_display_type_chooser = (self.display_layer.display_data_channel
+                                          and self.display_layer.display_data_channel.data_item
+                                          and self.display_layer.display_data_channel.data_item.is_data_complex_type)
+        self.display_type_items = [_("Log Absolute"), _("Absolute"), _("Phase"), _("Real"), _("Imaginary")]
+        self._display_type_flags = ["log-absolute", "absolute", "phase", "real", "imaginary"]
+        self._display_type_reverse_map = {p: i for i, p in enumerate(self._display_type_flags)}
+
+        current_flag = display_data_channel.complex_display_type if display_data_channel and display_data_channel.complex_display_type else str()
+        self.current_display_type_index_model = Model.PropertyModel[int](self._display_type_reverse_map.get(current_flag, 0))
+
+        self.__data_index_listener = self.data_index_model.property_changed_event.listen(
+            ReferenceCounting.weak_partial(LinePlotDisplayLayerModel.__on_data_index_changed, self))
+        self.__display_type_index_listener_model = self.current_display_type_index_model.property_changed_event.listen(
+            ReferenceCounting.weak_partial(LinePlotDisplayLayerModel.__on_display_type_index_changed, self))
+        self.__display_layer_list_added_listener = display_layers_list_model.item_added_event.listen(
+            ReferenceCounting.weak_partial(LinePlotDisplayLayerModel.__on_display_layer_list_changed, self))
+        self.__display_layer_list_removed_listener = display_layers_list_model.item_removed_event.listen(
+            ReferenceCounting.weak_partial(LinePlotDisplayLayerModel.__on_display_layer_list_changed, self))
 
-        def change_data_index(data_index_widget: UserInterface.LineEditWidget, display_layer: DisplayItem.DisplayLayer, data_index_str: str) -> None:
-            data_index = Converter.IntegerToStringConverter(pass_none=True).convert_back(data_index_str)
-            display_data_channel = display_item.display_data_channels[data_index] if data_index is not None else None
+    @property
+    def index(self) -> int:
+        if self.display_layer in self.display_item.display_layers:
+            return self.display_item.display_layers.index(self.display_layer)
+        else:
+            # we occasionally get an old LinePlotDisplayLayerModel that is not in the list
+            return 0
```

Comment:
This should never happen and indicates a bug somewhere else. How do you reproduce this?

Reply:
I added a listener to detect display layers being added/removed to trigger a property changed event for the index so the "Layer x" text could update. When I tested that with moving layers up/down the line `return self.display_item.display_layers.index(self.display_layer)` was throwing an exception saying the display_layer was not in the list. This didn't happen every time so I assumed it was a race condition where the old model was still attached to the event some of the time.